### ISSUE TITLE
Add kfirst

### ISSUE THREAD

**Issue by phadej** [well-typed/generics-sop#27]
```haskell
kfirst :: forall a b (x :: k). (a -> b) -> K a x -> K b x
```

`Bifunctor`'s `first` would restrict kind of `x`, it would be handy to have polykinded `kfirst`.

**Comment by kosmikus** [comment]
Not overly happy about the name (I'd prefer something with `map` in it, I think), but I'm certainly not opposed to adding it.

**Comment by kosmikus** [comment]
If I knew a good naming scheme, I would have already added the following functions:
```
ik :: (a -> x) -> I a -> K x b
iik :: (a -> b -> x) -> I a -> I b -> K x c
...
kk :: (x -> y) -> K x a -> K y b
kkk :: (x -> y -> z) -> K x a -> K y b -> K z c
...
ii :: (a -> b) -> I a -> I b
iii :: (a -> b -> c) -> I a -> I b -> I c
...
```
And possibly even more combinations.

**Comment by phadej** [comment]
hm, I don't think that
```haskell
iiMap :: _ -> I a   -> I b
ikMap :: _ -> I a   -> K b x
kiMap :: _ -> K a x -> I b
kkMap :: _ -> K a x -> K b y

iiiMap :: _ -> I a -> I b   -> I c
iikMap :: _ -> I a -> I b   -> K c x
ikiMap :: _ -> I a -> K b x -> I c
ikkMap :: _ -> I a -> K b x -> K c y

kiiMap :: _ -> K a x -> I b   -> I c
kikMap :: _ -> K a x -> I b   -> K c y
kkiMap :: _ -> K a x -> K b y -> I c
kkkMap :: _ -> K a x -> K b y -> K c z
```

is that bad idea, at least there is some logic behind those names. Reminds me lisp's `caddadr`...

**Comment by kosmikus** [comment]
I think we should probably do this as well. I'd suggest a variant of your naming scheme, @phadej, namely
```
mapII
mapIK
...
```

Has the advantage that the functor names remain uppercase.

I think these can just be added to Generics.SOP.BasicFunctors, and probably also re-exported via Generics.SOP. Up to how many arguments? 2?

**Comment by phadej** [comment]
Up to two arguments is fine; IIRC we provide other combinators only up to `hcliftA3` and `hczipWith3` anyway. Upper case variants are :+1:.

I could do PR tomorrow.

**Comment by kosmikus** [comment]
@phadej I've already started on this.

**Comment by kosmikus** [comment]
Fixed via #35.
